util/src/main/java/io/kubernetes/client/util/generic/GenericKubernetesApi.java

```diff
@@ -789,30 +1277,161 @@ KubernetesApiResponse<DataType> getKubernetesApiResponse(
-      }
-      if (null == status) { // the response body can be something unexpected sometimes..
-        // this line should never reach?
-        throw new RuntimeException(e);
-      }
-      return new KubernetesApiResponse<>(status, e.getCode());
+      return responseFromApiException(apiClient, e);
+    }
+  }
+
+  // For unit-test
+  interface CallAdapter extends UnaryOperator<Call> {}
+
+  private <DataType extends KubernetesType> Call prepareCall(
+      CallBuilder callBuilder, @Nonnull Consumer<KubernetesApiResponse<DataType>> callback)
+      throws ApiException {
+    Call call = prepareCall(callBuilder);
+
+    // For unit-test
+    if (callback instanceof CallAdapter) {
+      call = ((CallAdapter) callback).apply(call);
+    }
+
+    return call;
+  }
+
+  private <DataType extends KubernetesType>
+      Future<KubernetesApiResponse<DataType>> executeCallAsync(
+          ApiClient apiClient,
+          Class<DataType> dataClass,
+          CallBuilder callBuilder,
+          @Nonnull Consumer<KubernetesApiResponse<DataType>> callback) {
+    try {
+      Call call = prepareCall(callBuilder, callback);
+      CountDownLatch latch = new CountDownLatch(1);
+      AtomicReference<KubernetesApiResponse<DataType>> result = new AtomicReference<>();
+
+      apiClient.executeAsync(
+          call,
+          JsonElement.class,
+          new ApiCallback<JsonElement>() {
+            @Override
+            public void onFailure(
+                ApiException e, int statusCode, Map<String, List<String>> responseHeaders) {
+              if (latch.getCount() > 0) {
+                KubernetesApiResponse<DataType> r = responseFromApiException(apiClient, e);
+                result.set(r);
+                latch.countDown();
+                callback.accept(r);
+              }
+            }
+
+            @Override
+            public void onSuccess(
+                JsonElement element, int statusCode, Map<String, List<String>> responseHeaders) {
+              KubernetesApiResponse<DataType> r =
+                  getKubernetesApiResponse(dataClass, element, apiClient.getJSON().getGson());
+              if (latch.getCount() > 0) {
+                result.set(r);
+                latch.countDown();
+                callback.accept(r);
+              }
+            }
+
+            @Override
+            public void onUploadProgress(long bytesWritten, long contentLength, boolean done) {
+              // no-op
+            }
+
+            @Override
+            public void onDownloadProgress(long bytesRead, long contentLength, boolean done) {
+              // no-op
+            }
+          });
+
+      return new Future<KubernetesApiResponse<DataType>>() {
+        @Override
+        public boolean cancel(boolean mayInterruptIfRunning) {
+          if (isDone()) {
+            return false;
+          }
+          call.cancel();
+          boolean isCanceled = call.isCanceled();
```

Comment:
i'm not sure if `call#cancel()` is a synchronous/blocking process. it's not, then calling `call#isCancel()` right after `cancel()` will mostly always return false, wdyt?

Reply:
Oh, I understand what you are saying... My expectation is that `cancel()` is synchronous and that the Call is really canceled if it can be. The implementation of the concrete class is in Kotlin and looks like this:
```
  /**
   * Immediately closes the socket connection if it's currently held. Use this to interrupt an
   * in-flight request from any thread. It's the caller's responsibility to close the request body
   * and response body streams; otherwise resources may be leaked.
   *
   * This method is safe to be called concurrently, but provides limited guarantees. If a transport
   * layer connection has been established (such as a HTTP/2 stream) that is terminated. Otherwise
   * if a socket connection is being established, that is terminated.
   */
  override fun cancel() {
    if (canceled) return // Already canceled.

    canceled = true
    exchange?.cancel()
    connectionToCancel?.cancel()

    eventListener.canceled(this)
  }
```

Reply:
With this implementation, `isCanceled()` will always return `true`. I was more concerned with catching the rare case (different implementation class) where the call could not be canceled rather than a race-condition with an asynchronous version of `cancel()`.

I don't feel that strongly about this, so if you prefer, I can just have the method return `true` once `cancel()` is called.

Reply:
thanks for clarifying the behavior of `RealCall#cancel()`. the current implementation should work if that is a synchronous function.